lib/edac/memory_types.go

```diff
@@ -0,0 +1,168 @@
+package edac
+
+import (
+	"fmt"
+	"time"
+)
+
+// Information sourced from:
+// https://github.com/torvalds/linux/blob/master/Documentation/ABI/testing/sysfs-devices-edac
+
+type MemoryInfo struct {
+	// Name is the name of memory controller which generated this
+	// MemoryInfo.
+	Name string
+
+	// SinceReset is how long has elapsed since the last counter
+	// reset. This can be used with the error counters to measure
+	// error rates.
+	SinceReset time.Duration `file:"seconds_since_reset"`
+
+	// Type is the type of memory controller that is being utilized.
+	Type string `file:"mc_name"`
+
+	// Size in megabytes, of memory that this instance of memory
+	// controller manages.
+	Size int64 `file:"size_mb"`
+
+	// Uncorrectable is the total count of uncorrectable errors that
+	// have occurred on this memory controller.
+	// If panic_on_ue is set this counter will not have a chance to
+	// increment, since EDAC will panic the system.
+	Uncorrectable int64 `file:"ue_count"`
+
+	// UncorrectableNoInfo is the number of UEs that have occurred
+	// with no information as to which DIMM slot is having errors.
+	UncorrectableNoInfo int64 `file:"ue_noinfo_count"`
+
+	// Correctable is the total count of correctable errors that have
+	// occurred on this memory controller. This count is very important
+	// to examine. Correctable errors provide early indications that a
+	// DIMM is beginning to fail. This count field should be monitored
+	// for non-zero values and report such information to the system
+	// administrator.
+	Correctable int64 `file:"ce_count"`
+
+	// CorrectableNoInfo is the number of correctable errors that have
+	// occurred with no information as to which DIMM slot is having errors.
+	// Memory is handicapped, but operational, yet no information is
+	// available to indicate which slot the failing memory is in. This
+	// count field should be also be monitored for non-zero values.
+	CorrectableNoInfo int64 `file:"ce_noinfo_count"`
+
+	// ScrubRate is the actual scrubbing rate in bytes per second.
+	//
+	// If configuration fails or memory scrubbing is not implemented,
+	// then the value will be -1.
+	ScrubRate int64 `file:"sdram_scrub_rate"`
+
+	// MaxLocation is the last available memory slot in this memory
+	// controller. It is used by userspace tools in order to display the
+	// memory filling layout.
+	MaxLocation string `file:"max_location"`
+}
+
+// HasErrors returns true if mi has any errors, false otherwise.
+func (mi MemoryInfo) HasErrors() bool {
+	switch {
+	case mi.Correctable != 0:
+		return true
+	case mi.CorrectableNoInfo != 0:
+		return true
+	case mi.Uncorrectable != 0:
+		return true
+	case mi.UncorrectableNoInfo != 0:
+		return true
```

Comment:
not sure if doing `return mi.Correctable+mi.CorrectableNoInfo+mi.Uncorrectable+mi.UncorrectableNoInfo > 0` is a better solution here or not

Reply:
Not sure if these values can ever be negative or not

Reply:
The documentation for some of these fields states "This count field should be monitored for non-zero values and report such information to the system administrator." This gives no guarantee that these are not negative

Reply:
https://www.kernel.org/doc/html/v4.12/admin-guide/ras.html

Reply:
hmm, yeah, I'll revert that back to the previous switch behaviour then

Reply:
Updated to fix Stringer comments, there was two and moved the sourced comment as that was for all types not just MemoryInfo.